File: src/escalonamentoProcessos/sjf.js
import { tempoDeChegadaMinimo, vazio } from "./processo.js"

function ordenaProntos(a, b) {
    if (a.tc < b.tc)
        return -1;
    if (a.tc > b.tc)
        return 1;
    if (a.p < b.p)
        return -1;
    if (a.p > b.p)
        return 1;
    if (a.id < b.id)
        return -1;
    if (a.id > b.id)
        return 1;
    return 0;
}

export function sjf(processos) {
    let n = processos.length
    let retorno = tempoDeChegadaMinimo(processos)

    let processosProntos = processos
    processosProntos.sort(ordenaProntos)
    let it = 0
    let index = 0
    let index_aux = 0

    for(let i = 0; i < n; i++){
        processosProntos[i] = processos[i]
        retorno += processos[i].te
    }

    let escalonado = []

    console.log(`prontos[ ] vazio:\t${vazio(processosProntos)}`)
    console.log(`inseridos prontos[1]:\t${processosProntos[1].id} ${processosProntos[1].tc} ${processosProntos[1].te} ${processosProntos[1].d} ${processosProntos[1].p}\n`)

    while (it != processos.length){
        let aux = processosProntos[it]
        console.log(`index: ${index}\n`)
        for(let i = index; i < index + aux.te; i++){
            escalonado[i] = aux;
            index_aux++;
        }
        index = index_aux;
        it++;
    }

    return escalonado
}

File: src/escalonamento_processos/sjf.c
//
// Created by kid-a on 20/05/2022.
//

void prepara_lista_prontos(Processo processos[],  Processo lista_prontos[], int n){
    Processo p[n];
    Processo processo_vazio;
    processo_vazio.te = 0;
    int sumRetorno = 0;
    int menor = 0;
    int tam = n;
    int it = 0;
    int pivo;

    for(int i = 0; i < n; i++){
        p[i] = processos[i];
    }

    lista_prontos[it] = p[it];
    p[it] = processo_vazio;
    for(int i = it; i < n - 1; i++){
        Processo aux;
        aux = p[i];
        p[i] = p[i + 1];
        p[i + 1] = aux;
    }
    tam--;
    sumRetorno = lista_prontos[0].te;

    while (tamanho(p, n) > 1){
        pivo = 0;
        for(int i = 1; i < tam; i++){
            if((p[pivo].te <= p[i].te && p[pivo].tc < sumRetorno)){
                menor = pivo;
            }else if((p[i].tc < sumRetorno)){
                pivo = i;
                menor = i;
            }
        }
        it++;
        sumRetorno += p[menor].te;
        lista_prontos[it] = p[menor];
        p[menor] = processo_vazio;
        for(int i = menor; i < n - 1; i++){
            Processo aux;
            aux = p[i];
            p[i] = p[i + 1];
            p[i + 1] = aux;
        }
        tam--;
    }

    if(tamanho(p, n) == 1){
        lista_prontos[it] = p[0];
        p[0] = processo_vazio;
    }

    printf("\nprocessos ordenados:\n");
    for(int i = 0; i < n; i++){
        printf("processo %d\n", lista_prontos[i].id);
    }
    printf("\n");
}

Escalonado sjf(Processo processos[], int n){
    int retorno = tempo_de_chegada_minimo(processos, n);
    Processo prontos[n];
    prepara_lista_prontos(processos, prontos, n);

    int it = 0;
    int index = 0;
    int index_aux = 0;

    for(int i = 0; i < n; i++){
        prontos[i] = processos[i];
        retorno += processos[i].te;
    }

    Escalonado escalonado;
    escalonado.size = retorno;

    /*
    for(int i = 0; i < n; i++){
        prontos[i] = NULL;
    }*/

    printf("prontos[ ] vazio:\t%d\n", vazio(prontos, n));
    printf("inseridos prontos[1]:\t%d %d %d %d %d\n", prontos[1].id, prontos[1].tc, prontos[1].te, prontos[1].d, prontos[1].p);

    while (it != 4){
        Processo aux = prontos[it];
        for(int i = index; i < index + aux.te; i++) {
            escalonado.processos[i] = aux;
            index_aux++;
        }
        index = index_aux;
        it++;
    }

    return escalonado;
}